src/test/txospenderindex_tests.cpp

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) 2017-2022 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <chainparams.h>
+#include <index/txospenderindex.h>
+#include <test/util/setup_common.h>
+#include <util/time.h>
+#include <validation.h>
+
+#include <boost/test/unit_test.hpp>
+
+BOOST_AUTO_TEST_SUITE(txospenderindex_tests)
+
+BOOST_FIXTURE_TEST_CASE(txospenderindex_initial_sync, TestChain100Setup)
+{
+    TxoSpenderIndex txospenderindex(interfaces::MakeChain(m_node), 1 << 20, true);
+    BOOST_REQUIRE(txospenderindex.Init());
+
+    // Mine blocks for coinbase maturity, so we can spend some coinbase outputs in the test.
+     for (int i = 0; i < 50; i++) {
+         std::vector<CMutableTransaction> no_txns;
+         CreateAndProcessBlock(no_txns, this->m_coinbase_txns[i]->vout[0].scriptPubKey);
+     }
+    std::vector<COutPoint> spent(10);
+    std::vector<CMutableTransaction> spender(spent.size());
+
+    for (size_t i = 0; i < spent.size(); i++) {
+        spent[i] = COutPoint(this->m_coinbase_txns[i]->GetHash(), 0);
+        spender[i].version = 1;
+        spender[i].vin.resize(1);
+        spender[i].vin[0].prevout.hash = spent[i].hash;
+        spender[i].vin[0].prevout.n = spent[i].n;
+        spender[i].vout.resize(1);
+        spender[i].vout[0].nValue = this->m_coinbase_txns[i]->GetValueOut();
+        spender[i].vout[0].scriptPubKey = this->m_coinbase_txns[i]->vout[0].scriptPubKey;
+
+        // Sign:
+        std::vector<unsigned char> vchSig;
+        const uint256 hash = SignatureHash(this->m_coinbase_txns[i]->vout[0].scriptPubKey, spender[i], 0, SIGHASH_ALL, 0, SigVersion::BASE);
+        coinbaseKey.Sign(hash, vchSig);
+        vchSig.push_back((unsigned char)SIGHASH_ALL);
+        spender[i].vin[0].scriptSig << vchSig;
+    }
+
+    CreateAndProcessBlock(spender, this->m_coinbase_txns[0]->vout[0].scriptPubKey);
+
+    // Transaction should not be found in the index before it is started.
+    for (const auto& outpoint : spent) {
+        BOOST_CHECK(!txospenderindex.FindSpender(outpoint).has_value());
+    }
+
+    // BlockUntilSyncedToCurrentChain should return false before txospenderindex is started.
+    BOOST_CHECK(!txospenderindex.BlockUntilSyncedToCurrentChain());
+
+    BOOST_REQUIRE(txospenderindex.StartBackgroundSync());
+
+    // Allow tx index to catch up with the block index.
+    constexpr auto timeout{10s};
+    const auto time_start{SteadyClock::now()};
+    while (!txospenderindex.BlockUntilSyncedToCurrentChain()) {
+        BOOST_REQUIRE(time_start + timeout > SteadyClock::now());
+        UninterruptibleSleep(std::chrono::milliseconds{100});
+    }
+    for (size_t i = 0; i < spent.size(); i++) {
+        auto txId{txospenderindex.FindSpender(spent[i])};
+        BOOST_CHECK(txId.has_value() && txId.value() == spender[i].GetHash());
```

Comment:
nit: Should also use either snake_case `tx_id` or preferably `txid` to be consistent with the `Txid` type.

Could rely on `.value()` [throwing exception](https://en.cppreference.com/w/cpp/utility/optional/value) if not set and avoid the naming altogether.

\+ Slightly better for error messages etc to use `BOOST_CHECK_EQUAL`.

```suggestion
        BOOST_CHECK_EQUAL(txospenderindex.FindSpender(spent[i]).value(), spender[i].GetHash());
```

Reply:
(
If support for printing `std::optional` is added as in [pending PR 16545](https://github.com/bitcoin/bitcoin/pull/16545/files#diff-d4a2fb26adedc27f16bd3778424fa94c473342a695b228220a1810119028be5bR49), one could even write:
```suggestion
        BOOST_CHECK_EQUAL(txospenderindex.FindSpender(spent[i]), spender[i].GetHash());
```
)